### ISSUE TITLE
[Feature request] Ignore *all* Pandera type hints. Look, just do it!

### ISSUE THREAD

**Issue by ulfaslakprecis** [beartype/beartype#227]
Thanks for making this library. It's feels very... native (like it should be a part of the standard library) 🙏.

**I'd love to be able to ignore certain type hints**. In my specific case I am ALSO using `pandera` for Pandas DataFrame schema validation, and `beartype` doesn't seem to roll with the `pandera.DataFrame` type. My code looks like this:

```Python
import beartype as beartype
import pandera as pa

@beartype
@pa.check_types
def my_function(df: pa.DataFrame[MySchema], a: list[str], b: float):
    pass
```

It throws the exception:

> File "<@beartype(myfile.my_function) at 0x16be1fca0>", line 32, in my_function beartype.roar.BeartypeCallHintParamViolation: @beartyped myfile.my_function() parameter df="            column1 ...  column2... violates type hint pandera.typing.pandas.DataFrame[myFile.MySchema], as <protocol "pandas.core.frame.DataFrame"> "            column1  ...  column2... not instance of <protocol "pandera.typing.pandas.DataFrame">.

My temporary workaround is:

```Python
@beartype
@pa.check_types
def my_function(df: pa.DataFrame[MySchema] | pd.DataFrame, a: list[str], b: float):
    pass
```

I don't like it because it requires some verbose commenting for the next developer to understand what the heck is going on (they'd basically have to run the code and remove the ` | pd.DataFrame` part to see that it causes an error (the one above), and then have an aha-moment... not super nice).

**I'd like this pattern better:**

```Python
@beartype(ignore=[pa.DataFrame])  # or `ignore=["df"]`.
@pa.check_types
def my_function(df: pa.DataFrame[MySchema], a: list[str], b: float):
    pass
```

Maybe the best thing here would be if somehow `beartype` could work with `pandera`. But in a way I prefer explicitly not check `pa.DataFrame` objects with `beartype`, to signal that they are handled by `pandera`.

Cheers ☕️

**Comment by leycec** [comment]
> It's feels very... native (like it should be a part of the standard library) :pray:.

This is the highest praise I have ever received. Take my tears of joy, @ulfaslakprecis.

This was also my initial motivation for @beartype. I was shocked (and still am) that the standard library lacked even rudimentary support for type-checking. What kind of timeline is this? Which diety(ies) are in charge of Planet Earth, anyway? Somebody's sleeping on the job. There are now literally *hundreds* of Python Enhancement Proposals (PEPs) devoted exclusively to type-checking – yet Python doesn't officially support runtime type-checking of even a single PEP? Really? Everybody just has to use unofficial third-party packages from ~~crazy~~ "unique" people who live in the woods like @leycec? How could this be? How did things come to this?

I... I don't know, honestly. It is weird. There are official solutions for everything *else* in the Python ecosystem. It's not like we don't like official things. We do! Python has an [official build system](https://github.com/pypa/hatch), an [official package manager](https://pip.pypa.io/en/stable/), an [official package repository](https://pypi.org/), and an [official static type-checker](https://mypy-lang.org). We've done everything else. But we can't seem to do this.

I shrug laconically and then sling another 100 lines of poorly tested code a day. :muscle: 

> I am ALSO using `pandera` for Pandas `DataFrame` schema validation

`pandera`. That's a *beautiful* name for a package. That's also something I have never heard of. Now I have! Let me just add that to our [infinitely big list of awesome stuff to support when @leycec is not exhausted and can feel his hands](https://beartype.readthedocs.io/en/latest/pep).

Sadly, the @beartype exception raised above suggests that `pandera` type hints are not intended to be type-checked at either runtime *or* static type-checking time **by standards-compliant type checkers.** If they were, a Pandas `DataFrame` would clearly satisfy a `pandera.DataFrame[...]` type hint. Right? But it doesn't. So...

I grepped about the `pandera` codebase. I couldn't help myself. It's like sticking a greasy chopstick into your eye. You know it's gonna hurt bad, but your best bud that you have a long-standing bromance with just bet you $40 USD that you didn't have the intestinal fortitude to do it. What I'm trying to say here is...

**`pandera` is balls insane.** Seriously. I've never seen *anything* like that codebase. These mad lads actually built [their own ad-hoc PEP-noncompliant runtime type-checker just for the sole purpose of runtime type-checking their own ad-hoc PEP-noncompliant type hints](https://github.com/unionai-oss/pandera/blob/76bfa4f8546a2ba7f0e988ac8e454c35eeaa927e/pandera/decorators.py#L525). They voluntarily chose the path of mass insanity rather than just doing the Right Thing™, which was also the Easy Thing™: **make type hints PEP-compliant.**

Like, seriously. It's *not* hard to write standard type hints. All they had to do was define a metaclass for their type hints that implements the `__instancecheck__()` dunder method. Bam! Instant PEP-compliance. If they'd just done that:

* Their type hints would instantly work as expected out-of-the-box with Python's thriving ecosystem of (*waitforit*) PEP-compliant type-checkers. This means @beartype, [typeguard](https://github.com/agronholm/typeguard), the whole shebang, *everything*.
* They wouldn't have had to [write their own crazy PEP-noncompliant `@pandera.check_types()` type-checker](https://github.com/unionai-oss/pandera/blob/76bfa4f8546a2ba7f0e988ac8e454c35eeaa927e/pandera/decorators.py#L525).

Instead, they walked their own way. Now, their type hints don't actually work with anything and are actively breaking other things. This is what happens when you ignore community standards. This is the nuttiest thing I've seen today – and I follow [/r/UFOs](https://www.reddit.com/r/UFOs/), people. There's nothing nuttier than that place! It's all shadow biospheres, gubment spiracies, and obsolete F-16's shooting sidewinder missiles at non-existent unidentified aerial phenomena over at [/r/UFOs](https://www.reddit.com/r/UFOs/)! <sup>1</sup>

But `pandera`... man. Ancient astronauts got nuthin' on `pandera`. That codebase just takes the nonsensical cake.

<sup>1. Every man has his vice. Okay... *several* vices.</sup>

## tl;dr

@beartype will probably *never* support `pandera`, now that I've grepped through their code. They hate standards. @beartype loves standards. Only one of these opinions is the correct opinion.

**Comment by leycec** [comment]
*...heh.* Looks like I ranted uselessly like a rabid raccoon above. Let me now do something useful.

> My temporary workaround is...

Excellent workaround! Clever, too. The type hint `pa.DataFrame[MySchema] | pd.DataFrame` simultaneously satisfies both the PEP-noncompliant `@panderas.check_types` type-checker *and* the PEP-compliant `@beartype` type-checker. That's good. That type hint *also* unambiguously matches only Pandas `DataFrame` instances. That's also good. That said...

> I don't like it because it requires some verbose commenting for the next developer to understand what the heck is going on...

That's bad. You're *absolutely* right. The technical debt on million-line codebases is crippling enough without assaulting the new guy with labyrinthian type hints straight out a Lovecraft novella about Eldritch QA Paradigms from beyond the gaseous void of CPython. Therefore, this...

```python
@beartype(ignore=[pa.DataFrame])
```

**...is fascinating.** That's an excellent suggestion! We bow before your accumulated acumen. :bowing_man: 

That said, the `ignore` parameter should probably an accept an iterable of the *exact* type hints to be ignored, right? Under the hood, I can sadly guarantee you that there's no relation between the input `pa.DataFrame` **type hint factory** and the output `pa.DataFrame[MySchema]` **type hint.** From @beartype's far-flung perspective, the APIs of those two objects probably have nothing to do with one another. Ergo, you'd want to instead pass:

```python
@beartype(ignore=[pa.DataFrame[MySchema]])
```

Moreover, this feature request could actually be considered a duplicate of @beartype's existing feature request #191. Once implemented, #191 would provide a **plugin API** for third-party PEP-noncompliant type hints like `pandera.typing`. Via that plugin API, you (*...or preferably `pandera` devs themselves*) could implement runtime type-checking support for `pandera.typing` type hints in @beartype. In fact, "support" in this context could just mean instructing @beartype **to globally ignore *all* `pandera.typing` type hints.**

Actual code or it didn't happen:

```python
# None of this exists, currently. But it should!
from beartype.plug import Sign, map_repr_prefix_to_sign, map_sign_to_hint_transform

# Singleton object uniquely identifying *ALL* "pandera.typing" type hints.
SignOfThePandera = Sign('pandera')

# Tell @beartype to internally map *ALL* "pandera.typing" type hints to this sign.
map_repr_prefix_to_sign(repr_prefix='pandera.typing.', sign=SignOfThePandera)

# Tell @beartype to ignore *ALL* "pandera.typing" type hints.
map_sign_to_hint_transform(sign=SignOfThePandera, hint_transform=lambda obj: object)
```

Four lines of code. Nobody needs to understand that code. Even I barely understand that code. Just believe me when I tell you ~~comfortable lies~~ that that would configure @beartype to globally ignore *all* `pandera.typing` type hints.

In theory, that would probably be the optimal solution – right? Since the crazy `@pandera.check_types()` type-checker already type-checks `pandera.typing` type hints, there's *no* need for @beartype to do so. But there's also *no* need to repeatedly tell @beartype to explicitly ignore those type hints on *each and every* single usage of @beartype throughout your entire codebase. Right? That gets kinda ugly kinda fast.

Global is the way. So say we all.

**Comment by ulfaslakprecis** [comment]
Whoooa. What a response 🎉! Lots to unpack here. I'll be concise.

> pandera is balls insane. Seriously. I've never seen anything like that codebase. These mad lads actually built [their own ad-hoc PEP-noncompliant runtime type-checker just for the sole purpose of runtime type-checking their own ad-hoc PEP-noncompliant type hints](https://github.com/unionai-oss/pandera/blob/76bfa4f8546a2ba7f0e988ac8e454c35eeaa927e/pandera/decorators.py#L525). They voluntarily chose the path of mass insanity rather than just doing the Right Thing™, which was also the Easy Thing™: make type hints PEP-compliant.

This is super useful insight. I almost feel like opening an issue on their repository and link this issue. Unpredictable chaos might ensue, but in if they make the required changes to become PEP-compliant, everybody wins! Call it lobbying.

> Actual code or it didn't happen:
> 
> ```Python
> # None of this exists, currently. But it should!
> from beartype.plug import Sign, map_repr_prefix_to_sign, map_sign_to_hint_transform
> 
> # Singleton object uniquely identifying *ALL* "pandera.typing" type hints.
> SignOfThePandera = Sign('pandera')
> 
> # Tell @beartype to internally map *ALL* "pandera.typing" type hints to this sign.
> map_repr_prefix_to_sign(repr_prefix='pandera.typing.', sign=SignOfThePandera)
> 
> # Tell @beartype to ignore *ALL* "pandera.typing" type hints.
> map_sign_to_hint_transform(sign=SignOfThePandera, hint_transform=lambda obj: object)
> ```
> Four lines of code. Nobody needs to understand that code. Even I barely understand that code. Just believe me when I tell you comfortable lies that that would configure https://github.com/beartype to globally ignore all pandera.typing type hints.
> 
> In theory, that would probably be the optimal solution – right? Since the crazy @pandera.check_types() type-checker already type-checks pandera.typing type hints, there's no need for https://github.com/beartype to do so. But there's also no need to repeatedly tell https://github.com/beartype to explicitly ignore those type hints on each and every single usage of https://github.com/beartype throughout your entire codebase. Right? That gets kinda ugly kinda fast.

Hey man I don't understand it either, but I see what you're doing and I like it. IDK what goes on under the hood, but maybe just something simple that wraps the more complicated logic?

```Python
from beartype.plug import set_ignore_type_hints

set_ignore_type_hints(["pandera.typing"])
```

I agree that the simple `@beartype` decorator without arguments is the prettiest solution.

**Comment by leycec** [comment]
*...heh.* Third time's a charm. Let's do Pandera ~~dirty~~ right this time!

@beartype commit 0430574 is the first in a commit chain partially resolving this. In theory, that commit suffices to instruct @beartype to **unconditionally ignore *all* Pandera type hints.** Right? That's just the right thing to do here. @beartype should just "work" out-of-the-box with Pandera. This isn't your fault, so you shouldn't have to be tasked with telling  @beartype to ignore Pandera. This *is* Pandera's fault, but they're in no position to refactor their non-standard type hints to comply with standards; that'll never happen. I know humans! They'll just double down and blame everyone else.

But that commit assumes Pandera type hints comply with at least *some* standards by implementing the `__repr__()` dunder method of their type hints to return a string prefixed by `pandera.typing.`: e.g.,

```pycon
# @beartype assumes this works. It probably doesn't. Y u violate sanity, Pandera?
>>> from pandera import Column, DataFrameSchema
>>> from pandera.typing import DataFrame as DataFrameType
>>> muh_schema = DataFrameSchema({'column1': Column(int)})
>>> repr(DataFrameType[muh_schema])
pandera.typing.DataFrameType[DataFrameSchema({'column1': Column(int)})]  # <-- @beartype expects this.
```

This had better work, Pandera. I swear to all the Canadian Gods!

Naturally, that's not gonna work. Pandera defies every standard. :face_exhaling: 

Nonetheless, I *love* feisty confrontations like this. @beartype will find a way to trivially detect and ignore self-destructive type hints like Pandera's. First, I should probably install Pandera. Then, I should probably test. Theory crafting only gets the gentleman GitHubber so far. Let's see if I do anything tomorrow.

## Last But Definitely not Least...

Thanks so much for playing along with me and @beartype, @ulfaslakprecis. Your patience is infinite; your humour, even more infinite. We'll get this sorted for you and your ML workflow in a jiffy. You deserve the best. You humoured me.

This is the last feature request slated for our upcoming `beartype 0.13.0` release. Go, Team :bear:! Go!

**Comment by ulfaslakprecis** [comment]
> That approach does prevent you from decorating classes with @beartype (if you care about that kinda thing). Thankfully, decorating classes with @beartype is mostly just convenient syntactic sugar. I've become addicted to that sugar rush, myself. I can't imagine manually decorating class methods anymore... Gimme that sweet QA automation, @beartype.

Nono it works for classes too! I totally tried it out 🤞.

> Oh, and [I successfully packaged Pandera (and all of its transitive dependencies) for Gentoo Linux](https://github.com/leycec/raiagent/blob/master/dev-python/pandera/pandera-9999.ebuild) – my platform of choice, because I am a masochist. I'm delighted to announce that https://github.com/beartype commit https://github.com/beartype/beartype/commit/0430574e15e596a7301406f280d49bf51281e18d appears to ignore all Pandera type hints as hoped. Now, I just need to exhaustively exercise this with unit tests. Of course, unit testing Pandas + Panderas smells like non-trivial. I didn't even do anything tonight! Yet exhaustion has robbed my hands of the will to go on.

Sweet! I'll be watching for PyPI releases!!

**Comment by ulfaslakprecis** [comment]
Just adding a note here. If the update *ignores* all `pandera` type hints, we may have situations where something that is not even a `pandas.DataFrame` (could e.g. be a `numpy.ndarray`) passes undetected into a function that expects a `pandas.DataFrame` (though the type hint says `pandera.typing.DataFrame`. Case in point 👇:

```Python
X: np.ndarray[np.float64] = np.random((10, 10))

@beartype
def my_func(df: pandera.typing.DataFrame[MySchme]):
    # ... logic that assumes `X` is a `pandas.DataFrame`.

my_func(X)  # <-- throws error?
```

It is always the case that whatever the schema, the DataFrame is in reality a `pandas.DataFrame`. But if `beartype` ignores Pandera typehints, won't we have situations like this where `X` that is not even a pandas DataFrame, may pass into the function without causing an exception?

If so, I suppose you should ignore by casting to `pandas.DataFrame`. Maybe easier said than done.

**Comment by leycec** [comment]
**Gulp.** You're *absolutely* right, yet again. Of course you are. You're @ulfaslakprecis.

This is all Pandera's fault, I say! By running with scissors and their own PEP-noncompliant runtime type-checker, they've muddied the waters and manufactured all of these bizarre edge cases that otherwise wouldn't exist. I see at least *four* prominent edge cases here – best exemplified by your inspirational minimal-length example (MLE):

1. The new guy **forgot to apply the `@pandera.check_types()` decorator:** e.g.,

```python
# Uh oh. Pandera type hints *WITHOUT* the Pandera type-checker. Abort! Abort!
@beartype
def my_func(df: pandera.typing.DataFrame[MyScheme]):
    # ... logic that assumes `X` is a `pandas.DataFrame`.
```

@beartype can handle this by detecting whether the decorated callable (i.e., the function that the `@beartype` decorator is type-checking) is itself a Pandera-specific type-checker produced by the `@pandera.check_types` decorator. If *not*, @beartype will need to... what? Raise an exception? Emit a non-fatal warning? Magically make everything work anyway by silently decorating the decorated callable by `@pandera.check_types` for you on your behalf?

I know nothing. I only know that my forehead vein is already throbbing.

2. The new guy **applied the `@pandera.check_types` decorator *after* the `@beartype` decorator:** e.g.,

```python
# Uh oh. Beartype got there first. All hands on deck. Panic! Panic!
@pandera.check_types
@beartype
def my_func(df: pandera.typing.DataFrame[MyScheme]):
    # ... logic that assumes `X` is a `pandas.DataFrame`.
```

That new guy, right? Dude just can't get it right. Superficially, this edge case *seems* different from the prior edge case. Pragmatically, they're pretty much identical from @beartype's perspective. @beartype has *no* way of knowing that `@pandera.check_types` is going to be run immediately after `@beartype` is run. All `@beartype` knows is that `@pandera.check_types` has yet to be applied to the passed callable.

If Pandera was PEP-compliant, @beartype would go out of its way to make this madness "just work." But Pandera is PEP-noncompliant. We don't feel inclined to accommodate their *Rebel without a Cause* lifestyle. Instead, @beartype will probably just raise an exception in these cases.

Not @beartype's problem. *Next!* :point_right: 

3. The new guy **used Pandera type hints outside of a callable context:** e.g.,

```python
import pandera
from beartype import beartype
from beartype.door import die_if_unbearable
from dataclasses import dataclasses

# This is why standards are important, people.
@beartype
@dataclasses
class ThanksFerNuthinPandera(object):
    no_way_this_works: pandera.typing.DataFrame[MyScheme]

# standards. Standards. STANDA– *ahem*
die_if_unbearable('Not a Pandas DataFrame', pandera.typing.DataFrame[MyScheme])
```

Gosh dang it, new guy! He means well, but ultimately destroys everything he touches.

We all see the issue with the above, I trust. From the little that I can tell, Pandera type hints are contextually valid **only when annotating a callable decorated by the `@pandera.check_types` decorator.** The corollary is that Pandera type hints are contextually invalid **in all other use cases** – including as dataclass type hints or statement-level type hints passed to [@beartype's DOOR API](https://beartype.readthedocs.io/en/latest/api_door).

Again, @beartype will probably just raise an exception in these cases as well.

4. The new guy **applied the `@pandera.check_types` decorator *before* the `@beartype` decorator:** e.g.,

```python
# Finally. New guy did something right, everyone. Come gather around.
@beartype
@pandera.check_types
def my_func(df: pandera.typing.DataFrame[MyScheme]):
    # ... logic that assumes `X` is a `pandas.DataFrame`.
```

This is really the only Pandera workflow that @beartype can reasonably support, I think. To make this easier on everyone:

* We'll write up a new FAQ entry on type-checking Pandas `DataFrame` objects with @beartype **+** Pandera.
* We'll recommend that everyone automate the above decorator chaining with a trivial one-liner resembling:

```python
# Use this everywhere instead of @beartype or @pandera.check_type.
def bearpanderatype(func_or_cls):
    return beartype(pandera.check_types(func_or_cls))

# Feel the power of automation hum through your keyboard. 
@bearpanderatype
def my_func(df: pandera.typing.DataFrame[MyScheme]):
    # ... logic that assumes `X` is a `pandas.DataFrame`.
```

**Comment by leycec** [comment]
**Resolved by 7c8dfe7.** If this was any package other than Pandera, @beartype wouldn't have gone to the extreme lengths that it just did to support this. But this was Pandera – *the* Pandas validation framework. We can't ignore them, however much we desperately want to. Their defiance of community standards is a lesson to us all.

Exhaustive unit tests which are fragile and will almost certainly explode on the next release of Pandera and Pandas say this passes. I believe them. Technically, I should probably document how to integrate @beartype and Pandera with a new FAQ entry. Pragmatically, I am tired.

Blessed sleep! Take me into your somnambulant embrace. :sleepy:

**Comment by ulfaslakprecis** [comment]
You're a true gentleman @leycec. Sleep well ✨

**Comment by leycec** [comment]
**Thanks so much,** @ulfaslakprecis. Because your home town is the international capital of [music that is relevant to my tastes](https://rateyourmusic.com/genre/melodic-death-metal), the support of the Nords means everything. Truly, we are all typing gentlemen. Even this weird emoji rock idol guy → :man_singer: 

When `beartype 0.13.0` lands (...*please Gods let it be next week!*), it'd be so great if you and your intrepid team of superior ML savants could:

* Throw @beartype at every `@pandera.check_types`-decorated callable in your codebase.
* Submit another *superb* issue if @beartype stumbles into any bear traps.

We've exhaustively tested this, but that means nothing. This was deep voodoo. I would be pleasantly stunned if @beartype worked perfectly as advertised here. I want to believe. :sweat_smile:
